File: carpool-service/src/main/java/com/tompang/carpool/carpool_service/query/dto/RouteDto.java
package com.tompang.carpool.carpool_service.query.dto;

import org.locationtech.jts.geom.Point;

import com.tompang.carpool.carpool_service.query.entity.EventualAddress;

public class RouteDto {
    static final class Location {
        public final double latitude;
        public final double longitude;
        public final String address;

        public Location(double latitude, double longitude, String address) {
            this.latitude = latitude;
            this.longitude = longitude;
            this.address = address;
        }
    }

    public final Location origin;
    public final Location destination;

    public RouteDto(Point origin, Point destination, EventualAddress originAddress, EventualAddress destinationAddress) {
        this.origin = new Location(origin.getX(), origin.getY(), originAddress.toString());
        this.destination = new Location(destination.getX(), destination.getY(), destinationAddress.toString());
    }
}

File: webapp-bff/src/backend/carpool/dto/route.dto.ts
import { Type } from "class-transformer";
import { IsNumber, IsString, ValidateNested } from "class-validator";

class LocationDto {
  @IsNumber()
  latitude: number;
  @IsNumber()
  longitude: number;
  @IsString()
  address: string;
}

export class RouteDto {
  @ValidateNested()
  @Type(() => LocationDto)
  origin: LocationDto;
  
  @ValidateNested()
  @Type(() => LocationDto)
  destination: LocationDto;
}